src/exporters/file/task.rs

```diff
@@ -0,0 +1,458 @@
+use crate::bounded_channel::BoundedReceiver;
+use crate::exporters::file::config::FileExporterConfig;
+use crate::exporters::file::json::JsonExporter;
+use crate::exporters::file::parquet::{LogRecordRow, MetricRow, ParquetExporter, SpanRow};
+use crate::exporters::file::{FileExporterError, Result};
+use crate::init::file_exporter::FileExporterFormat;
+use chrono::Utc;
+use opentelemetry_proto::tonic::logs::v1::ResourceLogs;
+use opentelemetry_proto::tonic::metrics::v1::ResourceMetrics;
+use opentelemetry_proto::tonic::trace::v1::ResourceSpans;
+use std::path::Path;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, info};
+
+/// Generic trait for file exporters that can handle different data formats
+trait TypedFileExporter {
+    /// The type used to represent span data in this exporter
+    type SpanData: Clone;
+    /// The type used to represent metric data in this exporter
+    type MetricData: Clone;
+    /// The type used to represent log data in this exporter
+    type LogData: Clone;
+
+    /// Convert ResourceSpans to the exporter's span data format
+    fn convert_spans(&self, resource_spans: &ResourceSpans) -> Result<Vec<Self::SpanData>>;
+    /// Convert ResourceMetrics to the exporter's metric data format
+    fn convert_metrics(&self, resource_metrics: &ResourceMetrics) -> Result<Vec<Self::MetricData>>;
+    /// Convert ResourceLogs to the exporter's log data format
+    fn convert_logs(&self, resource_logs: &ResourceLogs) -> Result<Vec<Self::LogData>>;
+
+    /// Export span data to a file
+    fn export_spans(&self, data: &[Self::SpanData], path: &Path) -> Result<()>;
+    /// Export metric data to a file
+    fn export_metrics(&self, data: &[Self::MetricData], path: &Path) -> Result<()>;
+    /// Export log data to a file
+    fn export_logs(&self, data: &[Self::LogData], path: &Path) -> Result<()>;
+
+    /// Get the file extension for this exporter
+    fn file_extension(&self) -> &'static str;
+}
```

Comment:
You may be able to make this a typed trait and then remove the need for listing out the individual types, so something like:
```
trait TypedFileExporter<Resource> {
  type Output;

  fn convert(&self, items: &Resource) -> Result<Vec<Self::Output>>;
  fn export(&self, &[Self::Output]) -> Result<()>;
}
```

Reply:
In theory that could then allow you to replace the run_metrics_loop, run_traces_loop, run_logs_loop below?

Reply:
An example to look at would be the TransformPayload trait that has implementation for each concrete type (ResourceLogs, metrics, etc):
https://github.com/streamfold/rotel/blob/752692d84d39c1c5d7978b48b13bd6fd64fe499e/src/exporters/clickhouse/transform_logs.rs#L13-L15